File: arrays/medium/sort012.cpp
#include <bits/stdc++.h>
using namespace std;

/*
    The brute force solution for this would be to use any of the one sorting algorithm. If we use Merge Sort thr Time complexity would be O(NlogN) with the Space Complexity of O(N).
*/


// this is the better solution 
// Time Complexity: O(2N)
// Space complexity: O(1)
// void sort012(vector<int> &arr){
//     int cnt0 = 0;
//     int cnt1 = 0;
//     int cnt2 = 0;
//     for(int i = 0; i < arr.size(); i++){
//         if(arr[i] == 0) cnt0++;
//         if(arr[i] == 1) cnt1++;
//         if(arr[i] == 2) cnt2++;
//     }
//     for(int i = 0; i < cnt0; i++){
//         arr[i] = 0;
//     }
//     for(int i = cnt0; i < cnt0+cnt1; i++){
//         arr[i] = 1;
//     }
//     for(int i = cnt0+cnt1; i < arr.size(); i++){
//         arr[i] = 2;
//     }
// }


// This is the Optimal solution with Time complexity of O(N) and space Complexity of O(1).
void sort012(vector<int> &arr){
    int n = arr.size();
    int low = 0;
    int mid = 0;
    int high = n-1;
    while(mid <= high){
        if(arr[mid] == 0){
            swap(arr[mid], arr[low]);
            mid++;
            low++;
        }
        else if(arr[mid] == 1){
            mid++;
        }
        else if(arr[mid] == 2){
            swap(arr[mid], arr[high]);
            high--;
        }
    }
}



int main(){
    vector<int> arr = {2,0,2,1,1,0};
    sort012(arr);
    for(auto el: arr){
        cout << el << " ";
    }
    return 0;
}

File: arrays/medium/sort012.py
# Brute force solution to sort an array containing only 0s, 1s, and 2s using merge sort.

def mergeSort012(arr):
    if len(arr) <= 1:
        return arr
    
    mid = len(arr) // 2
    left_half = mergeSort012(arr[:mid])
    right_half = mergeSort012(arr[mid:])
    return merge(left_half, right_half)


def merge(left, right):
    sorted_array = []
    i = j = 0
    
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            sorted_array.append(left[i])
            i += 1
        else:
            sorted_array.append(right[j])
            j += 1
            
    while i < len(left):
        sorted_array.append(left[i])
        i += 1
        
    while j < len(right):
        sorted_array.append(right[j])
        j += 1
        
    return sorted_array


result = mergeSort012([0, 1, 2, 0, 1, 2, 1, 2, 0, 0, 1])
print(result) 



# Better Solution: Keeping Counts of 0s, 1s, and 2s.
def betterSort012(arr):
    count0 = 0
    count1 = 0
    count2 = 0
    for i in arr:
        if i == 0: count0 += 1
        elif i == 1: count1 += 1
        elif i == 2: count2 += 1
    
    for i in range(count0): arr[i] = 0
    for i in range(count0, count0+count1): arr[i] = 1
    for i in range(count0+count1, len(arr)): arr[i] = 2
    return arr


resultBetter = betterSort012([0, 1, 2, 0, 1, 2, 1, 2, 0, 0, 1])
print(resultBetter)


def swap(arr, i, j):
    arr[i], arr[j] = arr[j], arr[i]

# Optimized Solution: Dutch National Flag Problem
def sort012(arr):
    low, mid, high = 0, 0, len(arr) - 1
    while mid <= high:
        if arr[mid] == 0:
            swap(arr, low, mid)
            low += 1
            mid += 1
        elif arr[mid] == 1:
            mid += 1
        elif arr[mid] == 2:
            swap(arr, mid, high)
            high -= 1
    return arr

resultOptimized = sort012([0, 1, 2, 0, 1, 2, 1, 2, 0, 0, 1])
print(resultOptimized)

        
